packages/completer/src/handler.ts

```diff
@@ -477,6 +511,98 @@ export namespace CompletionHandler {
     connector: IDataConnector<IReply, void, IRequest>;
   }
 
+  /**
+   * Wrapper object for ICompletionItem.
+   */
+  export interface ICompletionItems {
+    /**
+     * isIncomplete is true if this is a partial list of completion items.
+     */
+    isIncomplete: boolean;
+
+    /**
+     * Callback which can be used for actions such as fetching full documentation.
+     */
+    resolve?: (id: string) => void;
+
+    /**
+     * Collection of completion items.
+     */
+    items: Array<ICompletionItem>;
+  }
+
+  /**
+   * Completion item object based off of LSP CompletionItem
+   */
+  export interface ICompletionItem extends PartialJSONObject {
+    /**
+     * User facing completion.
+     * If insertText is not set, this will be inserted.
+     */
+    label: string;
+
+    /**
+     * Completion to be inserted.
+     */
+    insertText?: string;
+
+    /**
+     * Range to be replaced by this completion.
+     */
+    range?: IRange;
+
+    /**
+     * Type of this completion item.
+     */
+    type?: string;
+
+    /**
+     * Image url for icon to be rendered with completion type.
+     */
+    icon?: string;
+
+    /**
+     * A human-readable string with additional information
+     * about this item, like type or symbol information.
+     */
+    documentation?: string;
+
+    /**
+     * A number used to help sort a set of completion items.
+     */
+    score?: number;
```

Comment:
I'd vote for removing this, and just using the array ordering, but this may be contentious.

The one good use-case for a score is if eventually we want to support asynchronously adding completions into the completions list (so the full ordering is not known beforehand).

But we can always add fields to this interface later without breaking compatibility, so it's safe to not include it.

Reply:
If we do this, we just need to have the old API sort completions according to its criteria before sending it to the new API.

Reply:
I removed the `score` property from the `ICompleteItem` interface.

> If we do this, we just need to have the old API sort completions according to its criteria before sending it to the new API.

The flow now is:

1. Process old completions API (`IReply`) into `ICompletionItems`, and send that to the model. Set boolean in model `isLegacy` if this is the case.
2. [When the model is asked to provide `items()`](https://github.com/kiteco/jupyterlab/pull/1/files#diff-a87b8d7967bd482ef72fe8656b494987R166) we then call `_markup()` and `_sort()`.
  * `_markup()` uses a helper function `StringExt.matchSumOfSquares` to both help it determine if there's a prefix match, for visual highlighting purposes, and computes a score based on how many characters match.
  * We then add a `score` property in the `ICompleteModel` so that it can be leveraged by `_sort()`.

I could see a world where we run `StringExt.matchSumOfSquares` twice; first just for scoring and sorting purposes in the old API transformation step, and then again to markup the completions since the latter would always be desirable. 

Is that what you're suggesting as well?

Reply:
ah, that was not what I was suggesting, but that works!

I didn't consider that we're already passing `isLegacy`, so we can use that to flag behavior. My suggestion was that we'd sort before creating the `ICompletionItems`.